tests/unit/test_charm.py

```diff
@@ -82,761 +137,432 @@ def _read_file(path: str) -> str:
 
-        patch_nrf_url.return_value = VALID_NRF_URL
-        patch_check_output.return_value = POD_IP
-        self.harness.set_can_connect(container=self.container_name, val=True)
-        self._create_nrf_relation()
-        self._create_certificates_relation()
+        self.harness.set_can_connect(container=CONTAINER_NAME, val=True)
 
-        self.harness.container_pebble_ready(self.container_name)
-        patch_restart.assert_not_called()
+        self.harness.container_pebble_ready(CONTAINER_NAME)
+        self.mock_restart.assert_not_called()
 
-    @patch(f"{CERTIFICATES_LIB}.get_assigned_certificates")
-    @patch("charm.check_output")
-    @patch("charms.sdcore_nrf_k8s.v0.fiveg_nrf.NRFRequires.nrf_url", new_callable=PropertyMock)
-    @patch("ops.model.Container.restart")
     def test_config_pushed_but_content_changed_and_layer_already_applied_when_pebble_ready_then_nssf_service_is_restarted(  # noqa: E501
-        self, patch_restart, patch_nrf_url, patch_check_output, patch_get_assigned_certificates
+        self, add_storage, fiveg_nrf_relation_id, certificates_relation_id
     ):
-        self.harness.add_storage(storage_name="certs", attach=True)
-        self.harness.add_storage(storage_name="config", attach=True)
-
         provider_certificate = Mock(ProviderCertificate)
         provider_certificate.certificate = STORED_CERTIFICATE
         provider_certificate.csr = STORED_CSR.decode()
-        patch_get_assigned_certificates.return_value = [provider_certificate]
+        self.mock_get_assigned_certificates.return_value = [provider_certificate]
 
-        root = self.harness.get_filesystem_root(self.container_name)
+        root = self.harness.get_filesystem_root(CONTAINER_NAME)
         (root / CSR_PATH).write_text(STORED_CSR.decode())
         (root / CERT_PATH).write_text(STORED_CERTIFICATE)
         (root / CONFIG_PATH).write_text(self._read_file(EXPECTED_CONFIG_FILE_PATH))
 
-        patch_check_output.return_value = b"1.2.3.4"
-        self.harness.set_can_connect(container=self.container_name, val=True)
-        patch_nrf_url.return_value = VALID_NRF_URL
-        self._create_nrf_relation()
-        self._create_certificates_relation()
+        self.mock_check_output.return_value = b"1.2.3.4"
+        self.harness.set_can_connect(container=CONTAINER_NAME, val=True)
 
-        self.harness.container_pebble_ready(self.container_name)
+        self.harness.container_pebble_ready(CONTAINER_NAME)
         self.harness.evaluate_status()
-        self.assertEqual(self.harness.model.unit.status, ActiveStatus(""))
-        patch_restart.assert_called_once_with(self.container_name)
+        assert self.harness.model.unit.status == ActiveStatus("")
+        self.mock_restart.assert_called_once_with(CONTAINER_NAME)
 
-    @patch("charms.sdcore_nrf_k8s.v0.fiveg_nrf.NRFRequires.nrf_url", new_callable=PropertyMock)
     def test_given_cannot_connect_to_container_when_nrf_available_then_status_is_waiting(
```

Comment:
`test_given_cannot_connect_to_container_when_nrf_available_then_status_is_waiting` and `test_given_cannot_connect_to_container_when_certificates_relation_changed_then_status_is_waiting` are exactly the same